### PR TITLE
Idiomatic Error Handling (#6)

### LINKED ISSUE #6: Idiomatic Error Handling
Wrap `on_http_request_headers` and `on_http_request_body` with internal function to allow the `?`-operator.

### PR DESCRIPTION
**Pull request by antonengelhardt**
Closes #6: Implement own Error Struct `PluginError` to handle  errors idiomatically.

### CODE REVIEW

**Comment by antonengelhardt** [comment]
Merge before #15

File: src/lib.rs [hunk 1]
```diff
@@ -411,21 +381,13 @@ impl ConfiguredOidc {
...
+                        Err(e) => return Err(PluginError::Utf8Error(e))
                     }
                 }
                 // If no body is found, log the error
-                None => {
-                    return Err(format!(
-                        "No body in token response with invalid status code."
-                    ));
-                }
+                None => return Err(PluginError::NoBodyError)
```
Comment (miwig) [inline src/lib.rs:390]: You could map `None` to an `Err` and get rid of the whole `match` like this:
```rust
let body = self
    .get_http_call_response_body(0, body_size)
    .ok_or(PluginError::NoBodyError)?;
```

**Review by miwig — COMMENTED** [review]
I like this a lot better already, and I think it could be even better still. I think a match with `Err(e) => return Err(...)` is almost always better expressed with the `?` operator. There is also some `unwrap` still left. For example `let url = Url::parse(&format!("{}{}", host, path)).unwrap();` will crash if the host does not contain a `:`.

By the way, why are there so many unrelated formatting changes or shuffled around imports? Those make it harder to review the PR. That's why I love `rustfmt`, I never have this problem.

File: src/cookie.rs [hunk 2]
```diff
@@ -56,42 +57,27 @@ impl AuthorizationState {
...
+        let decoded_cookie = base64engine.decode(cookie.as_bytes())?;
 
         // Decrypt with cipher
         let decrypted_cookie = match cipher.decrypt(nonce, decoded_cookie.as_slice()) {
             Ok(s) => s,
-            Err(e) => {
-                warn!("The cookie could not be decrypted: {}", e);
-                return Err(e.to_string());
-            }
+            Err(e) => return Err(PluginError::DecryptionError(e))
```
[the remark this answers is not in the document]
Comment (antonengelhardt) [reply src/cookie.rs:80]: Screenshot-Code-004855@2x

 @miwig

**Comment by antonengelhardt** [comment]
> I like this a lot better already, and I think it could be even better still. I think a match with `Err(e) => return Err(...)` is almost always better expressed with the `?` operator. There is also some `unwrap` still left. For example `let url = Url::parse(&format!("{}{}", host, path)).unwrap();` will crash if the host does not contain a `:`.

Yes, working on the other parts. Just wanted to get early feedback in, so that i'm on the right path.

> By the way, why are there so many unrelated formatting changes or shuffled around imports? Those make it harder to review the PR. That's why I love `rustfmt`, I never have this problem.

I sorted them alphabetically for better overview.

@miwig

**Comment by antonengelhardt** [comment]
> For example let url = Url::parse(&format!("{}{}", host, path)).unwrap(); will crash if the host does not contain a :.

We would need inner functions for this because this logic is placed in the function where the return is an `Action` and not a `Result`. I havent done this yet, because i think it makes the code less readable and more confusing. @miwig

File: src/error.rs [hunk 3]
```diff
@@ -0,0 +1,42 @@
+// thiserror
+use thiserror::Error;
+
+/// Error type for the plugin
+#[derive(Error, Debug)]
+pub enum PluginError {
```
[the remark this answers is not in the document]
Comment (antonengelhardt) [reply src/error.rs:6]: Actually the 503 code is returned when the token could not be stored in the cookie or when the token request fails. I can checkout anyhow on the weekend after in the time after that or you give it a spin :) 🌴  @miwig

**Comment by miwig** [comment]
> > For example let url = Url::parse(&format!("{}{}", host, path)).unwrap(); will crash if the host does not contain a :.
> 
> We would need inner functions for this because this logic is placed in the function where the return is an `Action` and not a `Result`. I havent done this yet, because i think it makes the code less readable and more confusing. @miwig

Unwrap or not, this particular example is just buggy because it doesn't work if the host doesn't have a port specified.
I think
 ```rust
let scheme = self
    .get_http_request_header(":scheme")
    .unwrap_or("http".into());

let url = Url::parse(&format!("{}://{}{}", scheme, host, path)).unwrap();
```
should be the right fix.

**Review by miwig — APPROVED** [review]

**Pull request merged**